test/test_relation.py

```diff
@@ -0,0 +1,269 @@
+# Copyright 2021 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import unittest
+
+from ops.charm import CharmBase
+from ops.framework import StoredState
+from ops.relation import Provider, Consumer
+from ops.testing import Harness
+
+##########################
+# Data used by all tests #
+##########################
+
+# Used to populate metadata.yaml
+METADATA = {
+    'relation_name': 'service',
+    'interface_name': 'svc'
+}
+
+# Used to setup provder and cosumer data
+CONFIG = {
+    'relation_name': METADATA['relation_name'],
+    'service_type': 'TestService',
+    'service_version': '1.0.0'
+}
+
+# Template for Provider metadata.yaml
+PROVIDER_META = '''
+name: provider-charm
+provides:
+  {relation_name}:
+    interface: {interface_name}
+'''
+
+# Template for Consumer metadata.yaml
+CONSUMER_META = '''
+name: provider-charm
+requires:
+  {relation_name}:
+    interface: {interface_name}
+'''
+
+# Template for Provider/Consumer charm config.yaml
+CONFIG_YAML = '''
+options:
+  relation_name:
+    type: string
+    description: 'Relation name used for testing'
+    default: {relation_name}
+  service_type:
+    type: string
+    description: 'Service type name used for testing'
+    default: {service_type}
+  service_version:
+    type: string
+    description: 'Service version used for testing'
+    default: {service_version}
+'''
+
+
+class ProviderCharm(CharmBase):
+    """A Provider charm used for testing"""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args)
+        self.provider = Provider(self,
+                                 self.model.config['relation_name'],
+                                 self.provides)
+
+    def ready(self):
+        self.provider.ready()
+
+    def unready(self):
+        self.provider.unready()
+
+    @property
+    def provides(self):
+        return {'provides': {
+            self.model.config['service_type']:
+            self.model.config['service_version']}}
+
+
+class ConsumerCharm(CharmBase):
+    """A Consumer charm used for testing
+
+    This charm records all events it receives in its stored state.
+    """
+    stored = StoredState()
```

Comment:
I would change this line, with this one:

```python
_stored = StoredState()
```

Reply:
Done.